New dev-todo-mvvm-databinding branch

JoseAlcerreca opened the pull request:
This is the first MVVM sample and, ideally, future MVVM samples will follow the same architecture (RX, Agera) so let's make sure it's generic and clear enough.

This already went through high-level review in my repo, so treat this as a final code review.

TODO: 
Complete README

oldergod commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,149 @@
+        mTasksRepository = tasksRepository;
+    }
+
+    public void start(String taskId) {
```
Would it not be better for the presenter to handle the data retrieving?

Your architecture graph also does not show any relations between the view model and the repositories: https://github.com/googlesamples/android-architecture/wiki/images/mvp-databinding.png

JoseAlcerreca replied:
The correct diagram is

https://raw.githubusercontent.com/wiki/googlesamples/android-architecture/images/mvvm-databinding.png

There is no presenter in this MVVM variant.

JoseAlcerreca commented on todoapp/app/src/main/AndroidManifest.xml:
```diff
@@ -33,8 +33,8 @@
-        <activity android:name="com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailActivity" />
-        <activity android:name="com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskActivity" />
+        <activity android:name="com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailActivity" android:exported="true"/>
+        <activity android:name="com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskActivity" android:exported="true"/>
```
No, good catch! I was using that for development.

JoseAlcerreca commented on todoapp/app/src/main/res/layout/tasks_frag.xml:
```diff
@@ -13,26 +13,29 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    android:onRefresh="@{actionHandler}">
+    android:onRefresh="@{viewmodel}"
```
That would be making the view (in this case the layout) a bit smarter. To keep it dumb, I prefer to have that call to loadTasks in the ViewModel.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:
```diff
@@ -60,40 +55,47 @@ public AddEditTaskFragment() {
     public void onActivityCreated(Bundle savedInstanceState) {
         super.onActivityCreated(savedInstanceState);
 
+        // FAB setup
```
if you need comments, how about moving these into 2 methods: `setupFabButton` and `setupSnackBar` and then remove the comments.

JoseAlcerreca replied:
Good idea, done.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:
```diff
@@ -102,23 +104,9 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
+        mViewModel.snackBarText.removeOnPropertyChangedCallback(mSnackBarChangedCallback);
     }
 }
+
```
remove line

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskNavigator.java:
```diff
@@ -16,32 +16,8 @@
-        void saveTask(String title, String description);
-
-        void populateTask();
+interface AddEditTaskNavigator {
```
Would be nice to have a javadoc for this.

JoseAlcerreca replied:
Done and done in the rest of navigators.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,149 @@
+
+    public final ObservableField<String> snackBarText = new ObservableField<>();
+
+    private static final String TAG = "TaskViewModel";
```
the tag is never used.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,149 @@
+
+    @Bindable
+    public String getTitleForList() {
+        if (mTask == null) {
```
same as above.

JoseAlcerreca replied:
Don't see the "above". I guess it's about method ordering, so done.

florina-muntenescu replied:
I had network issues and it looks like the comment for `getTitle` wasn't added. 
For both `getTitle` and `getTitleForList`, the case when the task is null is not unit tested (or I couldn't find it). Also, would be nice to have the `"No data"` string from the resources, not hardcoded, even if this is an "unlikely" situation.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java:
```diff
@@ -0,0 +1,139 @@
+public class AddEditTaskViewModel extends BaseObservable implements
+        TasksDataSource.GetTaskCallback, SnackBarChangedCallback.SnackBarViewModel {
+
+    public final String TAG = "AddEditTaskViewModel";
```
tag is not used

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java:
```diff
@@ -0,0 +1,139 @@
+
+    public final String TAG = "AddEditTaskViewModel";
+
+    private final TasksRepository mTasksRepository;
```
would be nice to have an order of members:
first public then private members. would make things easier to read. but, your choice.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.java:
```diff
@@ -17,38 +17,84 @@
+    public final ObservableBoolean error = new ObservableBoolean(false);
+
+    @VisibleForTesting
+    public int mNumberOfActiveTasks = 0;
```
you can make the members visible for testing package protected, not public.

JoseAlcerreca replied:
Done

oldergod commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.java:
```diff
@@ -16,15 +16,27 @@
+import static com.example.android.architecture.blueprints.todoapp.tasks.TasksFilterType.ALL_TASKS;
 
 /**
  * Exposes the data to be used in the {@link TasksContract.View}.
```
`TasksContract.View` does not exist anymore

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskActivity.java:
```diff
@@ -31,13 +31,14 @@
 
-    private AddEditTaskPresenter mAddEditTaskPresenter;
+    public static final String SHOULD_LOAD_DATA_FROM_REPO_KEY = "SHOULD_LOAD_DATA_FROM_REPO_KEY";
+    private AddEditTaskViewModel mViewModel;
```
add a new line

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Called by the Data Binding library.
```
but this is also called by the fragment when the user goes to menu->delete task
We don't seem to put any java doc in general, but if you want to mention that this is specifically called also by the data binding library, modify the javadoc to include also the "delete task" menu button.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksActivity.java:
```diff
@@ -137,4 +139,40 @@ public boolean onNavigationItemSelected(MenuItem menuItem) {
+        // If a task was successfully added, show mSnackBar
+        if (AddEditTaskActivity.REQUEST_CODE == requestCode) {
+
+            switch (resultCode) {
```
Displaying a text in the snackbar is UI logic. 
We could have a method in the model `handleActivityResult(result code)` that has this entire switch. And then this can be unit tested. 
Right now the snackbar string display logic is not tested in neither UI or unit tests.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java:
```diff
@@ -296,17 +246,29 @@ public View getView(int i, View view, ViewGroup viewGroup) {
+            });
+            viewmodel.start(task.getId());
+
+            //binding.executePendingBindings();
```
remove comment

JoseAlcerreca replied:
Done

JoseAlcerreca commented on README.md:
```diff
@@ -1,147 +1,17 @@
-  * Layout files are used to bind data to UI elements
-  * Events are also bound with an action handler
-  * Data can be observed and set up to be updated automatically when needed
+<pre>"View" refers to the architectural concept, not to `Android.view.View.`.</pre>
```
Done

JoseAlcerreca commented on README.md:
```diff
@@ -1,147 +1,17 @@
 
-<img src="https://github.com/googlesamples/android-architecture/wiki/images/mvp-databinding.png" alt="Diagram"/>
+The ViewModel in MVVM is very similar to the Presenter in MVP. The main difference has to do with the communication between View and ViewModel/Presenter:
+ - MVVM: when the ViewModel is modified, the View is updated by a library or framework. There's no way to update the View directly from the ViewModel, as it doesn't have this reference.
```
Can you elaborate?

JoseAlcerreca commented on README.md:
```diff
@@ -1,147 +1,17 @@
-The Data Binding library takes care of the communication between some
-components, so developers need to understand what it does and doesn't before
-making changes to the code.
+Observable fields in the ViewModel are *bound* to the specific TextViews, ImageViews, etc. via the layout files. The Data Binding library takes care of keeping View and ViewModel in sync. Bindings work in both ways: user actions automatically update the observable fields.
```
I'll wait for official docs for that.

JoseAlcerreca replied:
Done

oldergod replied:
Doesn't it exist already? https://developer.android.com/topic/libraries/data-binding/index.html

JoseAlcerreca commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/ScrollChildSwipeRefreshLayout.java:
```diff
@@ -52,4 +52,6 @@ public boolean canChildScrollUp() {
         mScrollUpChild = view;
     }
+
+
```
done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.java:
```diff
@@ -17,38 +17,84 @@
- * Exposes the data to be used in the {@link StatisticsContract.View}.
+ * Exposes the data to be used in the {@link StatisticsFragment}.
  * <p>
  * Note that in this case the view model is also the view, not the fragment.
```
The ViewModel is also the View? 
Shouldn't classes have only one responsibility? If you want to keep it like this it would be nice to add one more line explaining why do we have to make this a View.

JoseAlcerreca replied:
I don't know where that came from. Removed.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.java:
```diff
@@ -58,23 +53,35 @@ public static TaskDetailFragment newInstance(String taskId) {
 
-        mViewDataBinding.setPresenter(mPresenter);
+        // Set up floating action button
+        FloatingActionButton fab =
```
could you also wrap these in methods like you did in `AddEditTaskFragment`?

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class TaskDetailViewModel extends TaskViewModel {
+
+    private final TaskDetailNavigator taskDetailNavigator;
```
this is a private field so it should be `mTaskDetailsNavigator`

JoseAlcerreca replied:
Done.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TaskItemViewModel.java:
```diff
@@ -16,36 +16,37 @@
+        implements SnackBarChangedCallback.SnackBarViewModel {
 
-    private TasksContract.Presenter mListener;
+    private final TaskItemNavigator taskItemNavigator;
```
private field coding convention - `mTaskItemNavigator`

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java:
```diff
@@ -179,86 +138,77 @@ public boolean onMenuItemClick(MenuItem item) {
+        // Set up tasks view
+        ListView listView = mTasksFragBinding.tasksList;
+
+        // SnackBar setup
```
could you wrap all the setups in methods, like you did in the `AddEditTaskFragment`, please?

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksListBindings.java:
```diff
@@ -14,29 +14,29 @@
 
 /**
- * This specifies the contract between the view and the presenter.
+ * TODO: javadoc
```
[TODO](https://img.ifcdn.com/images/649ceade0fde522c86cd8d6a2f90be4a3c09146ec9211f68ab1f97fbb8b288ad_1.jpg) ?

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksListBindings.java:
```diff
@@ -14,29 +14,29 @@
+    public static void setItems(ListView listView, List<Task> items)
+    {
+        TasksFragment.TasksAdapter adapter = (TasksFragment.TasksAdapter) listView.getAdapter();
+        if (adapter != null)
```
Format the code.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.java:
```diff
@@ -16,39 +16,69 @@
+
+    private TasksFilterType mCurrentFiltering = TasksFilterType.ALL_TASKS;
+
+    public final ObservableList<Task> items = new ObservableArrayList<>();
```
Let's keep the order: public then private fields.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.java:
```diff
@@ -16,39 +16,69 @@
+    public final ObservableBoolean dataLoading = new ObservableBoolean(false);
 
-    private final TasksContract.Presenter mPresenter;
+    private final ObservableBoolean isDataLoadingError = new ObservableBoolean(false);
```
The [naming conventions](https://source.android.com/source/code-style.html#follow-field-naming-conventions) say that non-public field names start with "m".
So this means that we should have `mIsDataLoadingError` and `mSnackBarText`

JoseAlcerreca replied:
Done, snackbar text is package, so it's ok.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.java:
```diff
@@ -87,20 +117,117 @@ public Drawable getNoTaskIconRes() {
 
     @Bindable
     public boolean getTasksAddViewVisible() {
```
this method is not unit tested.

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.java:
```diff
@@ -87,20 +117,117 @@ public Drawable getNoTaskIconRes() {
+    }
+
+    /**
+     * Called by Data Binding library.
```
but also when clicking on the fab. Please update the doc

JoseAlcerreca replied:
Done

florina-muntenescu commented on todoapp/app/src/main/res/values/attrs.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright 2016, The Android Open Source Project
```
remove all the `~`

JoseAlcerreca replied:
That's IntelliJ's way, no need

florina-muntenescu commented on todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/tasks/TaskItemViewModelTest.java:
```diff
@@ -0,0 +1,134 @@
+    public void clickOnTask_ShowsDetailUi() {
+        loadTaskIntoViewModel();
+
+        mLoadTasksCallbackCaptor.getValue().onTaskLoaded(mTask); // Trigger callback
```
in all the other tests we put the comment above the line. 
Could we be consistent?

JoseAlcerreca replied:
They're different comments. The When..Then explains the test, whereas that one is a detail of what that particular line does, as arg captors are tricky.

JoseAlcerreca commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,146 @@
+
+
+/**
+ * TODO
```
Done

ravidsrk commented:
@JoseAlcerreca will this be merged?

JoseAlcerreca commented:
Yes.

yigit commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/SnackBarChangedCallback.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public void onPropertyChanged(Observable observable, int i) {
+        Snackbar.make(mView.get(),
```
this should first move the view to a local variable then check null on it before calling snakbar api.
Also, this whole class by itself seems kind of overkill.

JoseAlcerreca replied:
Done

yigit commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,146 @@
+/**
+ * Abstract class for View Models that expose a single {@link Task}.
+ */
+public abstract class TaskViewModel extends BaseObservable
```
i think you want to keep this class generic for other implementations but if not, it is better for it to have ObservableField, ObservableBoolean etc so that it won't need to deal w/ notifications.

JoseAlcerreca replied:
I've converted as many ObsFields as possible.

yigit commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,146 @@
+
+    private boolean mIsDataLoading;
+
+    public TaskViewModel(Context context, TasksRepository tasksRepository) {
```
seems like there is a responsibility conflict here.
This class is receiving the context but it is not defined what kind of context it is.
Because it is a ViewModel, i would guess it is the Activity context which means you cannot retain this ViewModel across config changes. This reduces the usability of this class.
I would strongly recommend keeping context (or almost anything android.*) out of this class (and make it jvm testable). Otherwise, this is more than a ViewModel.

yigit commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,146 @@
+            mTasksRepository.activateTask(mTask);
+            snackBarText.set(mContext.getResources().getString(R.string.task_marked_active));
+        }
+        notifyPropertyChanged(BR.completed);
```
again, getting rid of this custom field management would greatly simplify this code.

JoseAlcerreca replied:
But in this case, using two-way data binding, when the user marks a task as completed, the setCompleted() method is called. It's the perfect place to call the data repository and update the task status. 

If we create an ObservableBoolean I would have to add an OnPropertyChangedCallback. Do we prefer that?

yigit replied:
hmm you are right. i take it back.

yigit commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskActivity.java:
```diff
@@ -80,20 +82,19 @@ protected void onCreate(Bundle savedInstanceState) {
-        mAddEditTaskPresenter = new AddEditTaskPresenter(
-                taskId,
+        mViewModel = new AddEditTaskViewModel(
+                getApplicationContext(),
```
oh so above i thought this is created w/ activity context, turns out it is using application context. This is too confusing imo. It should just never receive context. And also, we should retain this ViewModel on configuration change via non-config state.

ErikHellman replied:
Crazy idea; how about making the ViewModel a singleton? The dependencies it has can be injected (and cleared) from the view at an appropriate life-cycle event (onCreate/onDestroy).

yigit replied:
if we make it singleton, it would be just a Model (app specific) rather than a ViewModel (screen specific). I think we are fine just if we retain it.

oldergod replied:
Why not passing the ViewModel between config changes with `onRetainCustomNonConfigurationInstance` ?

yigit commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskActivity.java:
```diff
@@ -80,20 +82,19 @@ protected void onCreate(Bundle savedInstanceState) {
     protected void onSaveInstanceState(Bundle outState) {
         // Save the state so that next time we know if we need to refresh data.
-        outState.putBoolean(SHOULD_LOAD_DATA_FROM_REPO_KEY, mAddEditTaskPresenter.isDataMissing());
+        outState.putBoolean(SHOULD_LOAD_DATA_FROM_REPO_KEY, mViewModel.dataLoading.get());
```
if we retain ViewModel, we can get rid of this code.

ErikHellman commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java:
```diff
@@ -0,0 +1,146 @@
+        mTask.setCompleted(isChecked);
+        if (isChecked) {
+            mTasksRepository.completeTask(mTask);
+            snackBarText.set(mContext.getResources().getString(R.string.task_marked_complete));
```
As far as I can tell, this is what we need the `Context` for in this class. 

I would change the `snackBarText` form an `ObservableField<String>` to an `ObservableInteger`. In the case where we would need more complex resource retrieval (string plurals etc), we can use an `ObservableField<Pair<Integer, Integer>>` or something to emit the value to the view. The view would then be responsible for the actual retrieval of the string resource.

ErikHellman commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java:
```diff
@@ -0,0 +1,137 @@
+public class AddEditTaskViewModel extends BaseObservable implements
+        TasksDataSource.GetTaskCallback, SnackBarChangedCallback.SnackBarViewModel {
+
+    public final ObservableField<String> title = new ObservableField<>();
```
Mixing `ObservableField/Boolean/XYZ` with `extends BaseObservable` feels a bit weird. I would go for using only `ObservableXYZ`.

JoseAlcerreca commented:
Addressed some of @yigit's comments. In summary: ViewModels are now retained (using non-UI Fragments) and I've converted some @Bindables to ObservableFields.

JoseAlcerreca commented:
Thanks everyone!

ravidsrk commented:
@JoseAlcerreca Please rename the branch to remove prefix **dev-** and also update the readme.
